mqtrigger/messageQueue/nats.go

```diff
@@ -107,35 +107,68 @@ func msgHandler(nats *Nats, trigger *crd.MessageQueueTrigger) func(*ns.Msg) {
 		log.Printf("Making HTTP request to %v", url)
 
 		headers := map[string]string{
-			"X-Fission-MQTrigger-Topic":     trigger.Spec.Topic,
-			"X-Fission-MQTrigger-RespTopic": trigger.Spec.ResponseTopic,
-			"Content-Type":                  trigger.Spec.ContentType,
+			"X-Fission-MQTrigger-Topic":      trigger.Spec.Topic,
+			"X-Fission-MQTrigger-RespTopic":  trigger.Spec.ResponseTopic,
+			"X-Fission-MQTrigger-ErrorTopic": trigger.Spec.ErrorTopic,
+			"Content-Type":                   trigger.Spec.ContentType,
 		}
 
 		// Create request
 		req, err := http.NewRequest("POST", url, bytes.NewReader(msg.Data))
+
+		if err != nil {
+			log.Errorf("Could not issue POST request with message to url %v", url)
+			return
+		}
+
 		for k, v := range headers {
 			req.Header.Add(k, v)
 		}
 
-		// Make the request
-		resp, err := http.DefaultClient.Do(req)
-		if err != nil {
-			log.Warningf("Request failed: %v", url)
+		var resp *http.Response
+		for attempt := -1; attempt < trigger.Spec.MaxRetries; attempt++ {
+			// Make the request
+			log.Warningf("Request : %v", req)
+			resp, err = http.DefaultClient.Do(req)
+			if err != nil || resp == nil {
+				// Retry without referencing status code of nil response on the next line
+				continue
+			}
+			if err == nil && resp.StatusCode == 200 {
+				// Success, quit retries
+				break
+			}
+		}
+
+		if resp == nil {
+			log.Warning("The response was undefined. Quit.")
 			return
 		}
+
+		// TODO : Sequence
 		defer resp.Body.Close()
 
-		body, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			log.Warningf("Request body error: %v", string(body))
+		body, bodyErr := ioutil.ReadAll(resp.Body)
+		if bodyErr != nil {
+			log.Warningf("Response body error: %v", bodyErr)
 			return
 		}
-		if resp.StatusCode != 200 {
-			log.Printf("Request returned failure: %v", resp.StatusCode)
-			return
+
+		// TODO : what if err is not nil, then body is empty. publish either of them, not just body.
```

Comment:
We shouldn't mix those errors into the same queue, simply avoiding publishing when body is empty is probably good enough.

Reply:
This comment was actually added by @smruthi2187. I'll remove this TODO and check if the body is empty before publishing.